### PR TITLE
Adding listener interfaces support and remove Netty from class names

### PR DESCRIPTION
**Pull request by isururanawaka**

### CODE REVIEW

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/internal/TransportServiceCapabilityProvider.java [hunk 1]
```diff
@@ -42,6 +41,6 @@ protected void start(BundleContext bundleContext) {
 
     @Override
     public int getCount() {
-        return YAMLTransportConfigurationBuilder.build().getListenerConfigurations().size();
+        return 1;
```
Comment (lankavitharana) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/internal/TransportServiceCapabilityProvider.java:45]: can we have a reason as a comment why we return 1?

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/CarbonHTTPServerInitializer.java [hunk 2]
```diff
@@ -95,15 +98,15 @@ public void setup(Map<String, String> parameters) {
                     log.debug("Disruptor is disabled and using executor thread pool with size of "
```
Comment (lankavitharana) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/CarbonHTTPServerInitializer.java:98]: better to have isDebugEnabled check

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        //Create Bootstrap Configuration from listener parameters
+        ServerBootstrapConfiguration.createBootStrapConfiguration(defaultListenerConfig.getParameters());
+        ServerBootstrapConfiguration serverBootstrapConfiguration = ServerBootstrapConfiguration.getInstance();
+        //boss group is for accepting channels
+        bossGroup = new NioEventLoopGroup(
+                bossGroupSize != 0 ? bossGroupSize : Runtime.getRuntime().availableProcessors());
+        //worker group is for processing IO
+        workerGroup = new NioEventLoopGroup(
+                workerGroupSize != 0 ? workerGroupSize : Runtime.getRuntime().availableProcessors() * 2);
+        log.debug("Netty Boss group size " + bossGroup);
```
Comment (lankavitharana) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java:117]: isDebugEnabled check

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        bootstrap = new ServerBootstrap();
+        bootstrap.option(ChannelOption.SO_BACKLOG, serverBootstrapConfiguration.getSoBackLog());
+        log.debug("Netty Server Socket BACKLOG " + serverBootstrapConfiguration.getSoBackLog());
+
+        bootstrap.group(bossGroup, workerGroup).channel(NioServerSocketChannel.class);
+
+        addChannelInitializer();
+
+        bootstrap.childOption(ChannelOption.TCP_NODELAY, serverBootstrapConfiguration.isTcpNoDelay());
+        log.debug("Netty Server Socket TCP_NODELAY " + serverBootstrapConfiguration.isTcpNoDelay());
```
Comment (lankavitharana) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java:128]: isDebugEnabled check

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+                if (artifactDeployer != null) {
+                    artifactDeployer.registerTransportListener(this);
+                }
+                log.info("HTTP Listener starting on port " + defaultListenerConfig.getPort());
+            } else {
+                log.error("HTTP Listener cannot start on port " + defaultListenerConfig.getPort());
+            }
+
+        } catch (InterruptedException e) {
+            log.error("HTTP Listener cannot start on port " + defaultListenerConfig.getPort(), e);
```
Comment (lankavitharana) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java:158]: can we log e.getMessage as well?

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/test/java/org/wso2/carbon/transport/http/netty/util/TestUtil.java [hunk 6]
```diff
@@ -69,15 +71,20 @@ public static void cleanUp(NettyListener nettyListener, HTTPServer httpServer) {
...
+                interfacesSet);
+        TransportSender transportSender = new HTTPSender(senderConfiguration);
+        HTTPTransportContextHolder.getInstance().setMessageProcessor(carbonMessageProcessor);
         carbonMessageProcessor.setTransportSender(transportSender);
         Thread transportRunner = new Thread(() -> {
             try {
-                nettyListener.start();
+                httpTransportListener.start();
             } catch (Exception e) {
                 LOGGER.error("Unable to start Netty Listener ", e);
```
Comment (lankavitharana) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/test/java/org/wso2/carbon/transport/http/netty/util/TestUtil.java:89]: can we append e.getMessage as well?

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/common/Constants.java [hunk 7]
```diff
@@ -89,6 +89,12 @@
 
     public static final String SERVER_BOOTSTRAP_SO_TIMEOUT = "server.bootstrap.socket.timeout";
 
+    public static final String SERVER_BOOTSTRAP_BOSS_GROUP_SIZE = "server.bootstrap.boss.group.size";
```
Comment (ramith) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/common/Constants.java:93]: why not document the usage of these constants?

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/internal/TransportServiceCapabilityProvider.java (hunk 1 above)
Comment (ramith) [reply http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/internal/TransportServiceCapabilityProvider.java:45]: +1

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/CarbonHTTPServerInitializer.java [hunk 8]
```diff
@@ -40,20 +40,23 @@
...
+public class CarbonHTTPServerInitializer extends ChannelInitializer<SocketChannel>
         implements CarbonTransportInitializer {
 
-    private static final Logger log = LoggerFactory.getLogger(CarbonNettyServerInitializer.class);
+    private static final Logger log = LoggerFactory.getLogger(CarbonHTTPServerInitializer.class);
     private ConnectionManager connectionManager;
 
-    private ListenerConfiguration listenerConfiguration;
+    private ListenerConfiguration defaultListenerConfiguration;
+    private Map<String, ListenerConfiguration> listenerConfigurationMap;
```
Comment (ramith) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/CarbonHTTPServerInitializer.java:50]: explaining these variables (usage, why they exists) adds value to code.

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+    private static final Logger log = LoggerFactory.getLogger(HTTPTransportListener.class);
+
+    private EventLoopGroup bossGroup;
+    private EventLoopGroup workerGroup;
+
+    private int bossGroupSize;
+    private int workerGroupSize;
+
+    private ServerBootstrap bootstrap;
+    private ListenerConfiguration defaultListenerConfig;
```
Comment (ramith) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java:59]: Please explain these variables (usage etc)  in a java-doc ( future readers, 2-5 years down the line will appreciate that a lot)

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+    private Map<Integer, ChannelFuture> channelFutureMap = new ConcurrentHashMap<>();
+    private Map<String, ListenerConfiguration> listenerConfigurationMap = new HashMap<>();
+    private Map<String, ListenerConfiguration> listenerConfigMapWithHostPort = new HashMap<>();
+
+    private Map<String, SSLConfig> sslConfigMap = new ConcurrentHashMap<>();
+
+    public HTTPTransportListener(int bossGroupSize, int workerGroupSize,
```
Comment (ramith) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java:68]: java docs ?

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+            Set<ListenerConfiguration> listenerConfigurationSet) {
+        super("ServerBootStrapper");
+        if (listenerConfigurationSet.size() == 0) {
```
Comment (ramith) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java]: isEmpty()?

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+        this.workerGroupSize = workerGroupSize;
+    }
+
+    @Override
+    public void start() {
+        log.info("Starting  HTTP Transport Listener");
+        startTransport();
+    }
+
+    private void startTransport() {
```
Comment (ramith) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java:107]: No explanation on the magic you are doing here? :D

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+            } else {
+                log.error("HTTP Listener cannot start on port " + defaultListenerConfig.getPort());
+            }
+
+        } catch (InterruptedException e) {
+            log.error("HTTP Listener cannot start on port " + defaultListenerConfig.getPort(), e);
+        }
+    }
+
+    //Channel Initializer is responsible for create channel pipeline
```
Comment (ramith) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java:162]: how about making this a standard java doc?

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/message/HTTPCarbonMessage.java [hunk 14]
```diff
@@ -36,9 +36,9 @@
 /**
  * Netty based implementation for Carbon Message.
  */
-public class NettyCarbonMessage extends CarbonMessage {
+public class HTTPCarbonMessage extends CarbonMessage {
```
Comment (ramith) [inline http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/message/HTTPCarbonMessage.java:39]: don't you think with the name change above class comment should also change?

File: http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java (hunk 13 above)
Comment (chanakaudaya) [reply http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java:162]: Private method descriptions should not be exposed as javadoc.

Comment (ramith) [reply http/netty/components/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPTransportListener.java:162]: http://grepcode.com/file_/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/ArrayList.java/?v=source

**Pull request merged**